src/Core/Content/Test/Seo/SeoUrlGeneratorTest.php

```diff
@@ -66,7 +66,7 @@ protected function setUp(): void
             $this->getContainer()->get('router.default'),
             $this->getContainer()->get('request_stack'),
             $this->getContainer()->get('shopware.seo_url.twig'),
-            $this->getContainer()->get(TwigVariableParser::class)
+            $this->getContainer()->get('Shopware\Core\Framework\Adapter\Twig\SeoUrlTwigVariableParser')
```

Comment:
```suggestion
            $this->getContainer()->get(SeoUrlTwigVariableParser::class)
```
(and adding the correct use statement).

Reply:
🤔 SeoUrlTwigVariableParser::class is not a class. Its a copy of TwigVariableParser without copy of code.

Reply:
Ah sorry you are right, I did not read the service XML declaration careful enough :-) sorry.

But maybe to avoid confusion, it would be better to than rename the service name to something which does not look like a class. But tbh I do not really know what the preferred way would be, so I would wait for someone qualified to comment on this :-)

Reply:
i'm with @aragon999 as it might be quite confusing, also it is easy to make the same mistake in another place where we might add the `TwigVariableParser` in the future and use a different twig environment :thinking: 

What do you think of removing the `TwigVariableParser` in general from the DI container and using a factory instead, so all places that currently use the `TwigVariableParser` don't inject the parser directly from the container, but inject the factory, and then pass the used twig environment to the factory to get a concrete instance of the parser, this assumes that we always already depend on the concrete twig environment where we use the  `TwigVariableParser`, but i guess that should always be the case.

Of course removing the `TwigVariableParser` declaration from the DI container would be a breaking change, so we should rather deprecate the service declaration for the 6.6 version and make sure that we internally don't use it anymore, to make sure that issues like this won't be repeated in the future.

What do you think?

Reply:
A TwigVariableParserFactory sound's good to me. I have two build two of them. One for Argument `twig` and one for  `shopware.seo_url.twig` in service declaration. Or is there a more elegant way that I don't see right now?

Reply:
My idea would be to have one generic factory:
```php
class TwigVariableParserFactory
{
    public function getParser(Envritonment $twig): TwigVariableParser
    {
        return new TwigVariableParser($twig);
    }
}
```

And then inject that factory as is into the services and not the the parser directly, so the service constructor may look like this:
```php
class MyService
{
    private TwigVariableParser $parser;

    public function __construct(private Environment $twig, ParserFactory $factory)
    {
         $this->parser = $factory->getParser($twig);
    }
}
```

so there is no way to get the parser directly from the DI container, as it is easy to use the wrong parser